corehq/apps/sso/tasks.py

```diff
@@ -99,3 +105,50 @@ def send_idp_cert_expires_reminder_emails(num_days):
                 f"Failed to send cert reminder email for IdP {idp}: {exc!s}",
                 exc_info=True,
             )
+
+
+@task(bind=True, default_retry_delay=15 * 60, max_retries=10, acks_late=True)
+def update_sso_user_api_key_expiration_dates(self, identity_provider_id):
+    idp = IdentityProvider.objects.get(id=identity_provider_id)
+    enforce_key_expiration_for_idp(idp)
+
+
+def enforce_key_expiration_for_idp(idp):
+    num_updated = 0
+
+    if idp.max_days_until_user_api_key_expiration is None:
+        return num_updated
+
+    user_iter = get_users_for_email_domains(idp.get_email_domains())
+    max_expiration_date = datetime.datetime.utcnow() + datetime.timedelta(
+        days=idp.max_days_until_user_api_key_expiration)
+    key_iter = get_keys_expiring_after(user_iter, max_expiration_date)
+
+    for key_chunk in chunked(key_iter, 500):
+        batch = []
+        for key in key_chunk:
+            key.expiration_date = max_expiration_date
+            batch.append(key)
+        HQApiKey.objects.bulk_update(batch, ['expiration_date'])
+        num_updated += len(batch)
+
+    return num_updated
+
+
+def get_users_for_email_domains(domains):
+    user_db = router.db_for_read(User)
+    user_batch_size = 500
+
+    for domain in domains:
+        query = Q(username__endswith=domain)
```

Comment:
Does `domain` include the `@` symbol? If not, this is an ambiguous query for multiple domains with the same ending characters in the domain (like `domain.com` for  `user@domain.com` and `user@predomain.com`).
```suggestion
        query = Q(username__endswith='@' + domain)
```

Reply:
Fixed in 0bf1ee44c812f2238ce0b4d8d6fa77e8b7fd2d76